src/main/java/fi/aalto/cs/apluscourses/presentation/CourseSelectionViewModel.java

```diff
@@ -0,0 +1,37 @@
+package fi.aalto.cs.apluscourses.presentation;
+
+import fi.aalto.cs.apluscourses.utils.observable.ObservableProperty;
+import fi.aalto.cs.apluscourses.utils.observable.ObservableReadWriteProperty;
+import fi.aalto.cs.apluscourses.utils.observable.ValidationError;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+
+public class CourseSelectionViewModel {
+
+  private List<CourseItemViewModel> courses;
+
+  public final ObservableProperty<String> selectedCourseUrl
+      = new ObservableReadWriteProperty<>("", CourseSelectionViewModel::validateCourseUrl);
+
+  public CourseSelectionViewModel(List<CourseItemViewModel> courses) {
+    this.courses = courses;
+  }
+
+  public CourseItemViewModel[] getCourses() {
+    return courses.toArray(new CourseItemViewModel[0]);
+  }
+
+  private static ValidationError validateCourseUrl(String courseUrl) {
+    if (courseUrl.trim().isEmpty()) {
+      return () -> "Select a course";
+    }
+    try {
+      new URL(courseUrl);
+      return null;
```

Comment:
can you explain a bit why such an approach is chosen: why `null`, why `return` instead of `throw`, why `ValidationError` is an error (but not an exception at all)?

Reply:
also, I'm curious what would happen if a) `URL` is incorrect, b) `URL` is formally (syntactically) correct, but there's no course config file at "that end"?

Reply:
> can you explain a bit why such an approach is chosen: why null, why return instead of throw, why ValidationError is an error (but not an exception at all)?

This reuses the validation approach that `OurDialogWrapper` uses. @OlliKiljunen is probably better than me at explaining the design decisions there - I personally like the consistency in that our dialogs all use the same validation mechanism.

> also, I'm curious what would happen if a) URL is incorrect, b) URL is formally (syntactically) correct, but there's no course config file at "that end"?

An incorrect URL doesn't pass the validation, so for a) the "OK" option of the dialog "refuses to work". For b), the dialog will happily accept the URL, but then a notification is shown that something is wrong with the "other end" of the URL. It would maybe be nice to check the course configuration already in the dialog, but that would be quite a heavy operation, and the course configuration is already checked in the course project action later, so this keeps the change smaller.

Reply:
This is about validating user input. `null` means no validation errors -> input is valid. It is customary not to use exceptions in user input validation (as its kinda expected to be invalid).

If URL is incorrect this returns non-null value causing validation error message to be shown to the user. Whether or not URL points to a correct document is out of this class' scope as that can only be known after the request over network is made and at that time, the form this view model is associated with, is already gone.

Reply:
fiddling with the error messages a bit more I see it works nicely:

```log
13.21	The A+ Courses plugin encountered a network error: Please check your network connection and try again. Error message: 404 Not Found.

13.23	Failed to parse the course configuration file
			The A+ Courses plugin for IntelliJ could not set up the target course, perhaps due to a mistake by the course staff.
			As a result, many features of the plugin won't work as expected. If the problem persists, please contact the course staff.
			Error message: Course configuration file should consist of a valid JSON object.`